v7/_posts/design/2018-09-26-$expandForV4.01.md

```diff
@@ -0,0 +1,161 @@
+---
+layout: post
+title: "6.2 Design doc for $expand in context-url for OData V4.01"
+description: "Design doc for $expand in context-url for OData V4.01"
+category: "6. Design"
+---
+
+# $expand in Context Url
+
+## Introduction
+In form of metadata, context url in response from OData-compliant service provides a way to describe the response payload, and is used as control information to facilitate response processing by the client. 
+
+Here is an example of a simple query request and resulting response containing context url:
+
+Request url: http://services.odata.org/V4/TripPinService/People('russellwhyte')?$select=FirstName,LastName
+
+Response:
+```
+{
+    "@odata.context": "http://services.odata.org/V4/TripPinService/$metadata#People(FirstName,LastName)/$entity",
+    "@odata.id": "http://services.odata.org/V4/TripPinService/People('russellwhyte')",
+    "@odata.etag": "W/\"08D62407F53DE9ED\"",
+    "@odata.editLink": "http://services.odata.org/V4/TripPinService/People('russellwhyte')",
+    "FirstName": "Russell",
+    "LastName": "Whyte"
+}
+```
+As shown above, @odata.context annotation specifies the context url, and provides the following machine-readable description for response data:
+- It is regarding an entity from the "People" entity set,
+- It contains two property values "FirstName", "LastName".
+
+## Expanded Entity Specification for OData V4.01
+In OData V4.01, notion for expanded entity and projected expanded entity [format](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_ExpandedEntity) is introduced for $expand clause in context url.
```

Comment:
Note that expanded nav props with nested selects were always included in the contextUrl.  The only change in 4.01 was that expanded nav props with no nested selects/expands are included with empty parens.  So more accurate would be:

OData V4.0 specifies that expanded entities with nested selects are included in the context Url as the name of the navigation property suffixed with the comma separated list of selected properties, enclosed in parens. OData V4.01 [format](http://docs.oasis-open.org/odata/odata/v4.01/cs01/part1-protocol/odata-v4.01-cs01-part1-protocol.html#sec_ExpandedEntity) specifies that, in the absence of any nested selects, the expanded navigation property appears suffixed with empty parens. This is distinct, and may appear in addition to, the un-suffixed navigation property name, which indicates that the navigation property appears in the $select list (indicating that the navigationLink should be included in the response).

Reply:
Updated as suggested. Thanks.

---
In reply to: [227443781](https://github.com/OData/odata.net/pull/1266#discussion_r227443781) [](ancestors = 227443781)